client/my-sites/plans-features-main/plan-notices.tsx

```diff
@@ -0,0 +1,136 @@
+import { useSelector } from 'react-redux';
+import MarketingMessage from 'calypso/components/marketing-message';
+import Notice from 'calypso/components/notice';
+import { getDiscountByName } from 'calypso/lib/discounts';
+import { getCurrentUserCurrencyCode } from 'calypso/state/currency-code/selectors';
+import isSiteAutomatedTransfer from 'calypso/state/selectors/is-site-automated-transfer';
+import {
+	getSitePlanSlug,
+	isCurrentUserCurrentPlanOwner,
+} from 'calypso/state/sites/plans/selectors';
+import { calculatePlanUpgradeCredits } from 'calypso/state/sites/plans/selectors/calculate-plan-upgrade-credits';
+import { isCurrentPlanPaid, isJetpackSite } from 'calypso/state/sites/selectors';
+
+function usePlanUpgradeCreditsDisplay(
+	siteId: number,
+	visiblePlanNames: string[] = []
+): {
+	creditsValue: number;
+	isPlanUpgradeCreditEligible: boolean;
+} {
+	const isSiteOnPaidPlan = !! useSelector( ( state ) => isCurrentPlanPaid( state, siteId ) );
+	const currentSitePlanSlug = useSelector( ( state ) => getSitePlanSlug( state, siteId ) );
+	const creditsValue = useSelector( ( state ) =>
+		calculatePlanUpgradeCredits( state, siteId, visiblePlanNames )
+	);
+	const isNotJetpackSiteOrIsAtomicSite = !! useSelector(
+		( state ) => ! isJetpackSite( state, siteId ) || isSiteAutomatedTransfer( state, siteId )
+	);
+
+	const isHigherPlanAvailable = function () {
+		const visiblePlansWithoutEnterprise = visiblePlanNames.filter(
+			( planName ) => planName !== PLAN_ENTERPRISE_GRID_WPCOM
+		);
+		const highestPlanName = visiblePlansWithoutEnterprise.pop();
+		return highestPlanName !== currentSitePlanSlug;
+	};
+
+	const isUpgradeEligibleSite =
+		isSiteOnPaidPlan && isNotJetpackSiteOrIsAtomicSite && isHigherPlanAvailable();
+
+	return {
+		creditsValue,
+		isPlanUpgradeCreditEligible: isUpgradeEligibleSite && creditsValue > 0,
+	};
+}
+
+export default function PlanNotices( {
+	siteId,
+	isInSignup,
+	visiblePlanNames = [],
+	discountInformation: { withDiscount, discountEndDate },
+}: {
+	visiblePlanNames: string[];
+	isInSignup: boolean;
+	siteId: number;
+	discountInformation: {
+		withDiscount: string;
+		discountEndDate: Date;
+	};
+} ) {
+	const translate = useTranslate();
+	const [ isNoticeDismissed, setIsNoticeDismissed ] = useState( false );
+	const currencyCode = useSelector( ( state ) => getCurrentUserCurrencyCode( state ) );
+	const canUserPurchasePlan = useSelector(
+		( state ) =>
+			! isCurrentPlanPaid( state, siteId ) || isCurrentUserCurrentPlanOwner( state, siteId )
+	);
+	const activeDiscount = getDiscountByName( withDiscount, discountEndDate );
+	const { creditsValue, isPlanUpgradeCreditEligible } = usePlanUpgradeCreditsDisplay(
+		siteId,
+		visiblePlanNames
+	);
+
+	if ( isNoticeDismissed ) {
+		return null;
+	} else if ( ! canUserPurchasePlan ) {
+		return (
+			<Notice
+				className="plan-features__notice"
+				showDismiss={ true }
+				onDismissClick={ () => setIsNoticeDismissed( true ) }
```

Comment:
I don't think we need to use useCallback here. The child <Notice />  is a fairly simple component and there is nothing much to optimise there and we do not have a reason to expect it growing in complexity since it's simply a small notice.

Also since `setIsNoticeDismissed` is simply **setting state** do we need to add an indirection by adding it into another handler function? 

I think having it this way is more readable at a glance. WDYT?

Reply:
> I don't think we need to use useCallback here. The child is a fairly simple component and there is nothing much to optimise there and we do not have a reason to expect it growing in complexity since it's simply a small notice.

👍 

> Also since setIsNoticeDismissed is simply setting state do we need to add an indirection by adding it into another handler function?

I thought about this too, but I'd still prefer a named function to reuse. The same definition is repeated three times if anything (currently three, could even grow).